### PR TITLE
feat: adds CommonMetadata to HelmRelease postrenderer

### PR DESCRIPTION
**Pull request by abhijith-darshan**
This PR addresses Issue #1217 to pass common labels and common annotations to `postRender` kustomize

Following the existing convention of `CommonMetadata` that is used in `kustomize-controller` to add labels and annotations, `CommonMetadata` is added to `HelmRelease` spec so that a common set of annotations and labels can be propagated to all resources in the helm chart without having to know the values schema of the chart

### CODE REVIEW

**Comment by abhijith-darshan** [comment]
@hiddeco any further recommendations for this PR?

File: api/v2/helmrelease_types.go [hunk 1]
```diff
@@ -185,6 +185,14 @@ type ValuesReference = meta.ValuesReference
 
 // Kustomize Helm PostRenderer specification.
 type Kustomize struct {
+	// Common labels to set on all resources.
+	// When duplicate labels are found across postRenderers, the last one wins.
+	// +optional
+	CommonLabels map[string]string `json:"commonLabels,omitempty"`
```
Comment (hiddeco) [inline api/v2/helmrelease_types.go]: I wonder if it might be better to expose the full options available in Kustomize, rather than giving the false impression this makes use of the deprecated `commonLabels` feature.

Comment (stefanprodan) [reply api/v2/helmrelease_types.go]: We have `commonMetadata` in Flux Kustomization that doesn’t alter the label selectors. I would go with this option instead of the deprecated commonLabels.

Comment (abhijith-darshan) [reply api/v2/helmrelease_types.go]: I saw this PR - https://github.com/fluxcd/kustomize-controller/pull/817

> The main difference to the Kustomize [commonLabels](https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/commonlabels/) and [commonAnnotations](https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/commonannotations/), is that the controller sets the labels and annotations only to the top level metadata field, without patching the Kubernetes Deployment spec.template or the Service spec.selector.

but in this case it would be useful to propagate the information to even the annotations or labels of the pods created by `Deployment` 

As this is not the use case of flux kustomize resource I would request to allow this option in the helm postRenderer.

How about I change this to `commonMetadata` format as suggested but expose further boolean option to `IncludeTemplates`?

Would that be acceptable?

Comment (stefanprodan) [reply api/v2/helmrelease_types.go]: Changing the label selectors will always result in a Helm upgrade failure with an immutable field error. I don't see why would we expose a way for people to shoot themselves in the foot.

Comment (abhijith-darshan) [reply api/v2/helmrelease_types.go]: @stefanprodan - sorry I don't mean Deployment's `.spec.selector.matchLabels` and you are right they are immutable...

I meant Deployment's `template.metadata.labels` which can be modified and that is what kustomize `IncludeTemplates` does.

and I did it this way -

```go
	if k.CommonLabels != nil {
		cfg.Labels = make([]kustypes.Label, 0)
		cfg.Labels = append(cfg.Labels, kustypes.Label{
			Pairs:            k.CommonLabels,
			IncludeSelectors: false,
			IncludeTemplates: true,
		})
	}
```

Comment (abhijith-darshan) [reply api/v2/helmrelease_types.go]: With this I did not face any issues in `HelmRelease`, so I was asking if this can be exposed in `postRenderer` spec

Also what is recommended here? to import the `commonMetadata` types from `kustomize_controller` or be independent of it?

**Comment by abhijith-darshan** [comment]
> @abhijith-darshan the `commonMetadata` should be added to `.spec`, same as we do for Flux Kustomization. Please copy the type instead of adding a dependency to kustomize-controller. As for the implementation, we already have code that adds labels to all resources to track the HelmRelease owner, we can extend that code to add the extra metadata.

As suggested - 

- `commonMetadata` from `.spec.postRenderers` to `.spec`.
- removed the dependency to `kustomize-controller` and copied over the types for `CommonMetadata`
- merged the logic of `origin_labels.go` into `common_renderer.go` 
- introduced `transformers.go` to re-use _kustomize builtins_ `LabelTransformer` and `AnnotationsTransformer`
- `BuildPostRenderers` gives precedence to `OriginLabels` rendering over `CommonMetadata` in case of label key collision (unlikely to happen but the edge case is covered)
- Updated tests but please let me know if I have missed something

File: internal/postrender/common_renderer_test.go [hunk 2]
```diff
@@ -0,0 +1,190 @@
+/*
+Copyright 2021 The Flux authors
```
Comment (stefanprodan) [inline internal/postrender/common_renderer_test.go]:
```suggestion
Copyright 2025 The Flux authors
```

File: internal/postrender/transformer.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
+package postrender
```
Comment (stefanprodan) [inline internal/postrender/transformer.go]: Copyright header is missing

**Review by stefanprodan — COMMENTED** [review]
Please add a section to the docs about common metadata before https://github.com/fluxcd/helm-controller/blob/main/docs/spec/v2/helmreleases.md#post-renderers

File: api/go.mod [hunk 4]
```diff
@@ -21,9 +21,9 @@ require (
 	github.com/modern-go/reflect2 v1.0.2 // indirect
 	github.com/spf13/pflag v1.0.6 // indirect
 	github.com/x448/float16 v0.8.4 // indirect
-	golang.org/x/net v0.39.0 // indirect
-	golang.org/x/text v0.24.0 // indirect
-	golang.org/x/tools v0.32.0 // indirect
+	golang.org/x/net v0.40.0 // indirect
```
Comment (stefanprodan) [inline api/go.mod]: Please undo all changes to go.mod and go.sum in the api and the main ones

Comment (abhijith-darshan) [reply api/go.mod]: done in [132abe9](https://github.com/fluxcd/helm-controller/pull/1223/commits/132abe998801a8b45128f364326c4be68e81a9a7)

File: internal/postrender/transformer.go (hunk 3 above)
Comment (abhijith-darshan) [reply internal/postrender/transformer.go]: Done in [8576fb5](https://github.com/fluxcd/helm-controller/pull/1223/commits/8576fb56608829bc648258dc9d90dfa0f2e73424)

**Comment by stefanprodan** [comment]
@abhijith-darshan can you please squash the commits
